Code follow MVC model: Guestbook MVC

asv-tungpt opened the pull request.

asv-tungpt committed:
Code follow MVC model: Guestbook MVC

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,29 @@
+
+
+class Guestbook():
+	@staticmethod
```
trong OOP hãy dùng @classmethod thay cho @staticmethod

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,29 @@
+		return ndb.Key('Guestbook', guestbook_name)
+
+	@staticmethod
+	def get_greetings(guestbook_name):
```
- nên dời vào class Greeting
- get_guestbook_key → get_lastest(guestbook_name, count = 10, force_new = false)
  - force_new = true : không xài cache

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -31,7 +30,7 @@ def get_context_data(self, **kwargs):
+			greeting = Greeting(parent=Guestbook.get_guestbook_key(guestbook_name))
 			if users.get_current_user():
 				greeting.author = users.get_current_user()
 			greeting.content = request.POST.get('content')
```
GreetingView thêm 
- force_new
- sau khi put thêm greeting thì force_new = true là được

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -31,7 +30,7 @@ def get_context_data(self, **kwargs):
 		if request.method == 'POST':
 			guestbook_name = request.POST.get('guestbook_name')
-			greeting = Greeting(parent=get_guestbook_key(guestbook_name))
+			greeting = Greeting(parent=Guestbook.get_guestbook_key(guestbook_name))
```
thao tác tạo mới 1 greeting từ 1 dictionary, lưu DB cũng nên đóng gói vào trong class Greeting

asv-tungpt committed:
Cấu trúc lại phù hợp với MVC models

- Thay @staticmethod bởi @classmethod
- Dời get_guestbook_key vào Greeting và đổi tên thành get_lastest
- Tạo propertise force_new để kiểm soát việc update memcache
- Đóng gói việc tạo mới một greeting vào trong function creat_greeting.

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,32 @@
+
+	@classmethod
+	def get_lastest(cls, guestbook_name=DEFAULT_GUESTBOOK_NAME):
+		return ndb.Key('Guestbook', guestbook_name)
```
đổi thành ndb.Key(cls, guestbook_name)
→ nên dùng class object, không dùng string

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,32 @@
+
+class Guestbook():
+	@classmethod
+	def get_greetings(cls, guestbook_name):
```
sry sai chính tả
- get_lastest sửa thành get_latest

đang hiểu sai thiết kế của method get_latest
- Guestbook .get_guestbook_key → Greeting.get_latest(guestbook_name, count = 10, force_new = false)
- việc query danh sách greeting nên để trong class Greeting 
- get_lastest : return 1 danh sách các greeting

phucnpm replied:
```
Greeting.get_latest(guestbook_name, count = 10, force_new = false) {
      if force_new = true thì không dùng cache
      else thử dùng cache
}

```

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -31,33 +31,29 @@ def get_context_data(self, **kwargs):
-			memcache.set('%s:greetings' % guestbook_name, greetings)
 			context = HttpResponseRedirect(
 				'/?' + urllib.urlencode({'guestbook_name': guestbook_name}))
+			GreetingView.force_new = True
```
hãy dùng self.force_new, ~~GreetingView.force_new~~

asv-tungpt committed:
Chỉnh sửa cho phù hợp hơn với cấu trúc MVC

- Greeting.get_latest(guestbook_name, count = 10, force_new = false)
- self.force_new

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -31,33 +32,29 @@ def get_context_data(self, **kwargs):
 		else:
 			context = HttpResponseRedirect('/')
+		if self.force_new:
+			GreetingView.update_memcache(self, guestbook_name, greeting)
```
nếu đã self.force_new = True thì có cần thiết update cache lúc này ?

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,38 @@
+		return greetings
+
+	@classmethod
+	def get_query_update_memcache(self, guestbook_name, count):
```
- không return về value thì đừng đặt là get_...
- là helper method, không cần public ra ngoài thì hãy thêm prefix _
  - Ex: _query_and_update_memcache( ... )

asv-tungpt committed:
Feature/

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,38 @@
+		return greetings
+
+	@classmethod
+	def prefix_query_update_memcache(self, guestbook_name, count):
```
hiểu sai rồi, 
hãy sửa thành  _query_and_update_memcache

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,38 @@
+
+	@classmethod
+	def prefix_query_update_memcache(self, guestbook_name, count):
+		greetings_query = Greeting.query(ancestor=ndb.Key(Greeting, guestbook_name)).order(
```
sửa thành 
- cls. query()
- ndb.Key(cls, guestbook_name)
- cls.date

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -31,33 +32,21 @@ def get_context_data(self, **kwargs):
 	def post(self, request):
 		if request.method == 'POST':
```
nếu kế thừa View thì không cần check request.method == POST 
bỏ đi

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -31,33 +32,21 @@ def get_context_data(self, **kwargs):
 		if request.method == 'POST':
 			guestbook_name = request.POST.get('guestbook_name')
-			greeting = Greeting(parent=get_guestbook_key(guestbook_name))
+			greeting = Greeting.creat_greeting(guestbook_name)
```
đưa các thao tác bên dưới vào class Greeting.put_from_dict(object_dict)

```
            greeting = Greeting.creat_greeting(guestbook_name)
            if users.get_current_user():
                greeting.author = users.get_current_user()
            greeting.content = request.POST.get('content')
            greeting.put()
```

phucnpm commented:
khi check comment trên github thì check trên Github 
https://github.com/asv-rak/python-training/pull/11

vì comment có thể dc update nội dung nhưng mail sẽ không thông báo phần update

phucnpm commented:
test lại chức năng sau khi hoàn tất việc thay đổi source

asv-tungpt committed:
Complete follow MVC model

Chỉnh sửa theo comment
Test lại chức năng của ứng

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +10,40 @@
+			logging.error('Memcache set failed.')
+		return greetings
+
+	def put_from_dict(greeting, content):
```
put_from_dict( ) :   hãy truyền vào 1 dictionary chứa đầy đủ thông tin của 1 greeting

phucnpm replied:
put_from_dict( ) hãy thiết kế là classmethod

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -1,18 +1,18 @@
 		guestbook_name = self.request.GET.get('guestbook_name', DEFAULT_GUESTBOOK_NAME)
-		greetings = get_greetings(guestbook_name)
+		greetings = Greeting.get_lastest(guestbook_name, 10, self.force_new)
+		self.set_force_new(False)
```
nếu là class method hãy gọi từ class object 
`self.__class__.set_force_new( )`

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -29,35 +29,20 @@ def get_context_data(self, **kwargs):
+		Greeting.put_from_dict(Greeting.creat_greeting(guestbook_name), request.POST.get('content'))
+		context = HttpResponseRedirect(
+			'/?' + urllib.urlencode({'guestbook_name': guestbook_name}))
+		self.set_force_new(True)
```
dùng 
`self.__class__.set_force_new( )`

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +10,40 @@
+		return Greeting(parent=Greeting.get_guestbook_key(guestbook_name))
+
+	@classmethod
+	def get_guestbook_key(cls, guestbook_name=DEFAULT_GUESTBOOK_NAME):
```
get_guestbook_key( ) không phải là action của Greeting hãy cho vào class Guestbook
- đổi thành Guestbook.get_key(cls, guestbook_name=DEFAULT_GUESTBOOK_NAME)

asv-tungpt committed:
Complete follow OOP

- Đưa những thứ linh tinh với greeting vào trong lớp
- chuyển hàm get_key vào trong lớp

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,40 @@
+	@classmethod
+	def get_lastest(cls, guestbook_name, count=10, force_new=False):
+		if(not force_new):
+			greetings = memcache.get('%s:greetings' % guestbook_name)
```
trường hợp không có cache thì sao ? hãy handle chỗ này

phucnpm commented on djangoguestbook/guestbook/views.py:
```diff
@@ -29,35 +29,21 @@ def get_context_data(self, **kwargs):
-		else:
-			context = HttpResponseRedirect('/')
+		guestbook_name = request.POST.get('guestbook_name')
+		dict = {'guestbook_name': guestbook_name, 'author': users.get_current_user(), 'content': request.POST.get('content')}
```
trình bày 1 dictionary theo format này nhé
https://github.com/asv-rak/python-training/blob/feature/mvc/djangoguestbook/guestbook/views.py#L22-L27

asv-tungpt committed:
Control Memcache and Format

phucnpm commented on djangoguestbook/guestbook/models.py:
```diff
@@ -7,13 +9,44 @@
+	def get_lastest(cls, guestbook_name, count=10, force_new=False):
+		if(not force_new):
+			greetings = memcache.get('%s:greetings' % guestbook_name)
+			if greetings is not None:
```
dùng `if greetings` thôi

asv-tungpt committed:
Rut gon dieu kien
